2048.py

```diff
@@ -0,0 +1,275 @@
+    return fields_with_zero_value
+
+def place_next_element_on_board(board):
+    if len(get_fields_with_zero_value(board)) > 0:
+        next_field = random.choice(get_fields_with_zero_value(board))
+        next_field.value = 2
+    else:
+        print("No more space on the board. You lost!")
+
+
+def check_if_similar_neighboring(board):
+    for row in board:
+        for index in range(0, len(row)-1):
+            if row[index].value == row[index + 1].value:
+                return True
+    for column in get_columns(board):
+        for index in range(0, len(column)-1):
+            if column[index].value == column[index + 1].value:
+                return True
+    return False
+
+
+
+def game_over(board):
+    if len(get_fields_with_zero_value(board)) > 0:
+        return False
+
+
+def check_if_won(board):
+    winning_number = 16
+    for row in board:
+        for field in row:
+            if field.value == winning_number:
+                return True
+    return False
+
+
+def get_temporary_board_values(board):
+    temporary_board_values = []
+    for row in board:
+        for field in row:
+            temporary_board_values.append(field.value)
+    return temporary_board_values
+
+def get_board_values(board):
+    board_values = []
+    for row in board:
+        for field in row:
+            board_values.append(field.value)
+    return board_values
+
+def validate_players_move(board):
+    temporary_values_board = get_temporary_board_values(board)
+    players_move = str(input("Next move: ").lower())
+    if players_move == "a":
+        move_left(board)
+        sum_left(board)
+        move_left(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
+            print("INVALID MOVE!")
+        else:
+            place_next_element_on_board(board)
+    elif players_move == "w":
+        move_up(board)
+        sum_up(board)
+        move_up(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
+            print("INVALID MOVE!")
+        else:
+            place_next_element_on_board(board)
+    elif players_move == "d":
+        move_right(board)
+        sum_right(board)
+        move_right(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
+            print("INVALID MOVE!")
+        else:
+            place_next_element_on_board(board)
+    elif players_move == "s":
+        move_down(board)
+        sum_down(board)
+        move_down(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
+            print("INVALID MOVE!")
+        else:
+            place_next_element_on_board(board)
+
+
+def game(number_of_rows, number_of_columns):
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
```

Comment:
with current logic this print for sure is not necessary! You can `print_board` firstly in `while` loop

Reply:
Corrected. Also added printing board after losing game in game_over method.